Migration doc (#308)
signalfx/splunk-otel-collector-chart: MIGRATION-FROM-SCK.md

# Migration from Splunk Connect for Kubernetes

## Assumptions

You are running Splunk Connect for Kubernetes (SCK) 1.4.9 and want to migrate to Splunk OpenTelemetry Collector for Kubernetes. Previous versions of the SCK are out of scope, but this guide still applies. You might encounter some version specific issues (such as missing/changed config options) but that shouldn't prevent you from proceeding with migration.

## Changes in components

SCK has 3 components/applications:

> 1. Logs, metrics and traces from a Kubernetes cluster (deployed as a DaemonSet)

Reviewer: Logs, metrics, and traces

> 2. Application to fetch cluster metrics from a kubernetes cluster (deployed as a deployment)

Reviewer: Kubernetes cluster

> 3. Application to fetch kubernetes objects metadata from a kubernetes cluster (deployed as a deployment)

Reviewer: Kubernetes cluster

> All SCK applications use Fluentd to work with logs, metrics and objects. Fluentd has significant performance issues when used to fetch logs in a Kubernetes cluster with pods that have very high throughput.

Reviewer: logs, metrics, and objects

Splunk OpenTelemetry Collector for Kubernetes provides significant performance improvements over the SCK through use of the OpenTelemetry Collector agent and native OpenTelemetry functionality for logs collection rather than Fluentd. You can learn more about the performance characteristics of this new application [Performance of native OpenTelemetry logs collection](https://github.com/signalfx/splunk-otel-collector-chart/blob/main/docs/advanced-configuration.md#performance-of-native-opentelemetry-logs-collection).

> 

Reviewer: See [Performance of native OpenTelemetry logs collection](https://github.com/signalfx/splunk-otel-collector-chart/blob/main/docs/advanced-configuration.md#performance-of-native-opentelemetry-logs-collection) to learn more about the performance characteristics of this new application.

Splunk OpenTelemetry Collector for Kubernetes has the following components and applications:

> 1. Splunk OpenTelemetry Collector Agent (`agent`) to fetch logs, metrics and traces from a kubernetes cluster (deployed as a Kubernetes DaemonSet)

Reviewer: logs, metrics, and traces

Also, capitalize Kubernetes.

2. Splunk OpenTelemetry Collector Cluster Receiver (`clusterReceiver`) to fetch metrics from a Kubernetes API (deployed as a Kubernetes 1-replica Deployment).
3. Optional Splunk OpenTelemetry Collector Gateway (`gateway`) to forward data through it to reduce load on Kubernetes API and apply additional processing (deployed as a Kubernetes Deployment).

> 

Reviewer: Removed the periods from lines 22 and 23 as these are not complete sentences.

There is no application available for fetching Kubernetes objects metadata from a Kubernetes cluster.

> ### Changes in logs, metrics, and objects

Reviewer: logs, metrics, and objects

Author: what do you mean here?

Author: nvm, deleting this whole table as suggested

| Logging | Metrics | Objects |
|---|---|---|

> | Redhat UBI docker images for our applications are no longer available, as we now use scratch images | The naming convention of the metrics used in Splunk OpenTelemetry Collector for Kubernetes has changed and there are minor differences in the names of the metrics | Not implemented or available |

Reviewer: Red Hat Universal Base Image (UBI) Docker

| AWS FireLens is not supported. | Changes in the number of metrics (fewer metrics available) and additional metrics for the OpenTelemetry Collector |  |
|  |  |  |

### Changes in logs in Splunk OpenTelemetry Connector for Kubernetes

> * Red Hat Universal Base Image (UBI) Docker images for our applications are no longer available, as we now use scratch images.

Reviewer: Red Hat UBI

* AWS FireLens is not supported.

### Changes in metrics in Splunk OpenTelemetry Connector for Kubernetes

* The naming convention of the metrics used in Splunk OpenTelemetry Collector for Kubernetes follows the OpenTelemetry specification and is different than SCK. You will observe minor differences in the names of the metrics.

> * Previously in SCK, you could get a large number of metrics from various APIs for kubernetes. However, in recent versions of kubernetes 1.18+, these API sources are disabled by default and there are fewer metrics available. The previous list of metrics can be found [Metrics Information](https://github.com/splunk/fluent-plugin-kubernetes-metrics/blob/master/metrics-information.md).

Reviewer: See [Metrics Information](https://github.com/splunk/fluent-plugin-kubernetes-metrics/blob/master/metrics-information.md) for the previous list of metrics.

| SCK | Yes | Yes | Yes |
| Splunk OpenTelemetry Collector for Kubernetes | Yes | Yes | No |

As shown in the table, you can acquire logs and metrics using Splunk OpenTelemetry Collector for Kubernetes. If you have objects currently deployed and need objects data, leave your current SCK objects deployment as-is (Helm command args --set="splunk-kubernetes-logging.enabled=false,splunk-kubernetes-metrics.enabled=false,splunk-kubernetes-objects.enabled=true"). This allows you to use only the SCK for objects data together with Splunk OpenTelemetry Collector for Kubernetes for logs and other telemetry data.

### Checkpoint translation

With migration to the Splunk OpenTelemetry Collector for Kubernetes, the underlying framework/agent being used has changed, so there needs to be a translation for checkpoint data so that the new OpenTelemetry agent can continue where Fluentd left off. All of this occurs automatically as an initContainer when you deploy the new Helm chart in your cluster for the first time. If you are not using a custom path for checkpoint for fluentd, it should just work with default configurations for this new Helm chart.

To migrate Fluentd's position files again:

> 1. Stop OpenTelemetry Daemonset by deleting deployed Helm chart.

Reviewer: Stop the OpenTelemetry DaemonSet.

2. Delete the OpenTelemetry checkpoint files in the ```"/var/addon/splunk/otel_pos/"``` directory from Kubernetes nodes.

> 3. Restart the new Helm chart Daemonset.

Reviewer: DaemonSet

#### Custom metadata

If you are using the `customMetadata` option to set a custom metadata field in your logs in SCK, you can use the `extraAttributes.custom` option in Splunk OpenTelemetry Collector for Kubernetes.

#### Custom annotations

If you are using the `customMetadataAnnotations` option to set custom annotation fields in your logs in SCK, you can use the `extraAttributes.fromAnnotations` option in Splunk OpenTelemetry Collector for Kubernetes.

#### Pod scheduling configurations

If you are using any pod scheduling operations such as `nodeSelector`, `affinity`, and `tolerations` in SCK, you can use the `nodeSelector`, `affinity` and `tolerations` options in Splunk OpenTelemetry Collector for Kubernetes.

#### Service accounts

If you are using the `serviceAccount` option to use your own service accounts in SCK, you can use the `serviceAccount` option in Splunk OpenTelemetry Collector for Kubernetes.

#### Secrets

If you are using the `secret` option to use your own service accounts in SCK, you can use the `secret` option in Splunk OpenTelemetry Collector for Kubernetes.

> #### Custom docker image and pull secrets

Reviewer: Custom Docker

## Changes in components

SCK has 3 components/applications:

1. Logs, metrics and traces from a Kubernetes cluster (deployed as a DaemonSet)
2. Application to fetch cluster metrics from a kubernetes cluster (deployed as a deployment)
3. Application to fetch kubernetes objects metadata from a kubernetes cluster (deployed as a deployment)

All SCK applications use Fluentd to work with logs, metrics and objects. Fluentd has significant performance issues when used to fetch logs in a Kubernetes cluster with pods that have very high throughput.

Splunk OpenTelemetry Collector for Kubernetes provides significant performance improvements over the SCK through use of the OpenTelemetry Collector agent and native OpenTelemetry functionality for logs collection rather than Fluentd. You can learn more about the performance characteristics of this new application [Performance of native OpenTelemetry logs collection](https://github.com/signalfx/splunk-otel-collector-chart/blob/main/docs/advanced-configuration.md#performance-of-native-opentelemetry-logs-collection).

Splunk OpenTelemetry Collector for Kubernetes has the following components and applications:

> 1. Application to fetch logs and traces from a kubernetes cluster (deployed as a DaemonSet)
> 2. Application to fetch metrics and objects from a kubernetes cluster (deployed as a DaemonSet).

Reviewer 2: ```suggestion
1. Splunk OpenTelemetry Collector Agent (`agent`) to fetch logs, metrics and traces from a kubernetes cluster (deployed as a Kubernetes DaemonSet)
2. Splunk OpenTelemetry Collector Cluster Receiver (`clusterReceiver`) to fetch metrics from a Kubernetes API (deployed as a Kubernetes 1-replica Deployment).
3. Optional Splunk OpenTelemetry Collector Gateway (`gateway`) to forward data through it for reduce load on Kubernetes API and apply additional processing (deployed as a Kubernetes Deployment).
```

There is no application available for fetching Kubernetes objects metadata from a Kubernetes cluster.

### Changes in logs, metrics, and objects

| Logging | Metrics | Objects |
|---|---|---|
| Redhat UBI docker images for our applications are no longer available, as we now use scratch images | The naming convention of the metrics used in Splunk OpenTelemetry Collector for Kubernetes has changed and there are minor differences in the names of the metrics | Not implemented or available |
| AWS FireLens is not supported. | Changes in the number of metrics (fewer metrics available) and additional metrics for the OpenTelemetry Collector |  |

> |  |  |  |

Reviewer 2: This section seems redundant as everything is repeated below. Up to you to keep it or not

Author: it is redundant. i am okay either way. I will leave it until someone from doc comments

Reviewer: Agree with Dmitrii. I would remove the table.

* system.paging.operations
* system.paging.usage
* system.processes.count
* system.processes.created
* up

### Changes in objects in Splunk OpenTelemetry Collector for Kubernetes

This chart does not support the collection of Kubernetes objects. This is future functionality. You can use SCK to collect other objects.

## Migration overview

### Migration options/matrix

The following table shows the options for migrating from SCK to Splunk OpenTelemetry Collector for Kubernetes:

| Method  | Logs | Metrics | Objects |
|---|---|---|---|
| SCK | Yes | Yes | Yes |
| Splunk OpenTelemetry Collector for Kubernetes | Yes | Yes | No |

As shown in the table, you can acquire logs and metrics using Splunk OpenTelemetry Collector for Kubernetes. If you have objects currently deployed and need objects data, leave your current SCK objects deployment as-is (Helm command args --set="splunk-kubernetes-logging.enabled=false,splunk-kubernetes-metrics.enabled=false,splunk-kubernetes-objects.enabled=true"). This allows you to use only the SCK for objects data together with Splunk OpenTelemetry Collector for Kubernetes for logs and other telemetry data.

### Checkpoint translation

> With migration to the Splunk OpenTelemetry Collector for Kubernetes, the underlying framework/agent being used has changed, so there needs to be a translation for checkpoint data so that the new OpenTelemetry agent can continue where Fluentd left off. All of this occurs automatically as an initContainer when you deploy the new Helm chart in your cluster for the first time. If you have properly configured the values.yaml file for the new Helm chart, it will continue working without interruption. Without proper configuration for migration, there is a possibility of either data duplication or data loss while migrating from SCK to Splunk OpenTelemetry Collector for Kubernetes.

Reviewer 2: > If you have properly configured the values.yaml file for the new Helm chart, it will continue working without interruption. Without proper configuration for migration, there is a possibility of either data duplication or data loss while migrating from SCK to Splunk OpenTelemetry Collector for Kubernetes.

This sounds somewhat confusing if we don't provide any guidance on how to properly configure that. I'd call out that configuration is required only if user has a custom path for checkpoints for fluentd, otherwise it should work OOTB.

To migrate Fluentd's position files again:

> 1. Delete the OpenTelemetry checkpoint files in the ```"/var/lib/otel_pos/"``` directory from Kubernetes nodes.

Reviewer 2: ```suggestion
1. Delete the OpenTelemetry checkpoint files in the ```"/var/addon/splunk/otel_pos/"``` directory from Kubernetes nodes.
```

> 2. Restart the new Helm chart Daemonet."

Reviewer 2: I'd say 
1. stop 
2. remove dir 
3. start

## Step 1: Preparing your values.yaml file for migration

Translate the values.yaml file from SCK to an appropriate format for Splunk OpenTelemetry Collector for Kubernetes. The following are the configurations for SCK and Splunk OpenTelemetry Collector for Kubernetes:

* [SCK configuration](https://github.com/splunk/splunk-connect-for-kubernetes/blob/develop/helm-chart/splunk-connect-for-kubernetes/values.yaml)
* [Splunk OpenTelemetry Collector for Kubernetes configuration](https://github.com/andrewy-splunk/splunk-otel-collector-chart/blob/main/helm-charts/splunk-otel-collector/values.yaml)

### Translating global/generic/splunk configurations from SCK to Splunk OpenTelemetry Collector for Kubernetes

#### Specifying your Splunk Platform and HTTP Event Collector (HEC) configuration

> You can combine the `host`, `port` and `protocol` options from SCK to use the `endpoint` option in Splunk OpenTelemetry Collector for Kubernetes.

Reviewer 2: ```suggestion
You can combine the `host`, `port` and `protocol` options from SCK to use the `splunkPlatform.endpoint` option in Splunk OpenTelemetry Collector for Kubernetes.
```

This option uses this format: "http://X.X.X.X:8088/services/collector" which is interpreted as "protocol://host:port/services/collector".

> If you are using the `clientCert`, `clientKey`, and `caFile` options from SCK, use the corresponding `clientCert`, `clientKey`, and `caFile` options in Splunk OpenTelemetry Collector for Kubernetes to specify your HEC certificate chain.

Reviewer 2: ```suggestion
If you are using the `clientCert`, `clientKey`, and `caFile` options from SCK, use the corresponding `clientCert`, `clientKey`, and `caFile` options under `splunkPlatform` in Splunk OpenTelemetry Collector for Kubernetes to specify your HEC certificate chain.
```

> If you are using the `insecureSSL` option from SCK, use the `insecureSkipVerify` option in Splunk OpenTelemetry Collector for Kubernetes to specify whether to verify the certificates on HEC.

Reviewer 2: ```suggestion
If you are using the `insecureSSL` option from SCK, use the `splunkPlatform.insecureSkipVerify` option in Splunk OpenTelemetry Collector for Kubernetes to specify whether to verify the certificates on HEC.
```

> If you are using the `indexName` option from SCK, use the index option in Splunk OpenTelemetry Collector for Kubernetes to specify which index you want to index data into

Reviewer 2: ```suggestion
If you are using the `indexName` option from SCK, use the `splunkPlatform.index` option in Splunk OpenTelemetry Collector for Kubernetes to specify which index you want to index data into
```

## Translating custom configurations from SCK to Splunk OpenTelemetry Collector for Kubernetes for logs

You can find all the configuration options used to upgrade in the values.yaml files linked above for both SCK and Splunk OpenTelemetry Collector for Kubernetes.

#### Using root user/permissions for accessing log files

> If using root user/permissions to access log files in SCK, set `runAsUser: 0` in the `securityContext` in Splunk OpenTelemetry Collector for Kubernetes.

Reviewer 2: This is the default. probably can be omitted

#### Container runtimes

If you configured SCK to explicitly set any of the following, you can do the same with Splunk OpenTelemetry Collector for Kubernetes:

> * Set the `containerRuntime` to a value from CRI-O, containerd, or Docker, depending on your runtime.

Reviewer 2: ```suggestion
* Set the `logsCollection.containers.containerRuntime` to a value from CRI-O, containerd, or Docker, depending on your runtime.
```

* Set the `path` option to the location of your logs on your nodes.

> * If using the `exclude_path` option from SCK, you can use the `excludePaths` option in Splunk OpenTelemetry Collector for Kubernetes to exclude any logs you want.

Reviewer 2: ```suggestion
* If using the `exclude_path` option from SCK, you can use the `logsCollection.containers.excludePaths` option in Splunk OpenTelemetry Collector for Kubernetes to exclude any logs you want.
```

#### Custom docker image and pull secrets

If you are using custom Docker images, tags, pull secrets, and pull policy in SCK, you can achieve the same using the `image` option in Splunk OpenTelemetry Collector for Kubernetes to specify the relevant configs for using docker images.

#### Limiting or increasing application resources

If you are using the `resources` option in SCK to limit/increase the CPU and memory usage in SCK, you can do the same using the `agent.resources` option in Splunk OpenTelemetry Collector for Kubernetes.

#### Extra files from host

For tailing files other than container or journald logs (that is, kube audit logs), configure `logsCollection.extraFileLogs` using this [filelog](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/receiver/filelogreceiver) receiver configuration.

[SCK values.yaml snippet]

```yaml
logs:
  kube-audit:
    from:
      file:
        path: /var/log/kube-apiserver-audit.log
    timestampExtraction:
      format: "%Y-%m-%dT%H:%M:%SZ"
    sourcetype: kube:apiserver-audit
```

[Splunk OpenTelemetry Collector for Kubernetes values.yaml snippet]

```yaml
logsCollection:
  extraFileLogs:

>   filelog/kube-audit:

Author: fix indentation

* **If you want to delete only logs and metrics**

  * Update the values.yaml file used to deploy SCK to disable objects and run the following command:

  * helm upgrade local-k8s -f your-values-file.yaml splunk/splunk-connect-for-kubernetes

* **If you want to delete only logs**

  * Update the values.yaml file used to deploy SCK to disable objects and run the following command:

  * helm upgrade local-k8s -f your-values-file.yaml splunk/splunk-connect-for-kubernetes

* **If you want to delete your entire SCK deployment (logs, objects and metrics), run the following command**

  * helm delete local-k8s

## Step 3: Installing Splunk OpenTelemetry Collector for Kubernetes

* See [How to install](https://github.com/signalfx/splunk-otel-collector-chart/blob/main/README.md#how-to-install) for the instructions.

## Step 4: Check data in Splunk

* Check the logs index to see if you are receiving logs from your Kubernetes cluster
  * index="Your logs index"
* Check the metrics index to see if you are receiving logs from your Kubernetes cluster
  * | mcatalog values(metric_name) WHERE index="Your metrics index"

> ## Step 5: Delete SCK Deployment if not done in Step 2

Reviewer 2: A header without content seems wrong. This one probably not necessary and can be removed?
